tests/unit/models/test_input.py

```diff
@@ -67,55 +63,55 @@ class TestPackageInput:
-        assert package.dict() == expect_data
+        adapter = pydantic.TypeAdapter(PackageInput)
+        package = cast(PackageInput, adapter.validate_python(input_data))
+        assert package.model_dump() == expect_data
 
     @pytest.mark.parametrize(
         "input_data, expect_error",
         [
             (
                 {},
-                r"Discriminator 'type' is missing",
+                r"Unable to extract tag using discriminator 'type'",
             ),
             (
                 {"type": "go-package"},
-                r"No match for discriminator 'type' and value 'go-package' \(allowed values: .*",
+                r"Input tag 'go-package' found using 'type' does not match any of the expected tags: 'gomod', 'npm', 'pip', 'yarn'",
             ),
             (
                 {"type": "gomod", "path": "/absolute"},
-                r"path\n  path must be relative: /absolute",
+                r"Value error, path must be relative: /absolute",
             ),
             (
                 {"type": "gomod", "path": ".."},
-                r"path\n  path contains ..: ..",
+                r"Value error, path contains ..: ..",
             ),
             (
                 {"type": "gomod", "path": "weird/../subpath"},
-                r"path\n  path contains ..: weird/../subpath",
+                r"Value error, path contains ..: weird/../subpath",
             ),
             (
                 {"type": "pip", "requirements_files": ["weird/../subpath"]},
-                r"requirements_files -> 0\n  path contains ..: weird/../subpath",
+                r"pip.requirements_files\n  Value error, path contains ..: weird/../subpath",
             ),
             (
                 {"type": "pip", "requirements_build_files": ["weird/../subpath"]},
-                r"requirements_build_files -> 0\n  path contains ..: weird/../subpath",
+                r"pip.requirements_build_files\n  Value error, path contains ..: weird/../subpath",
             ),
             (
                 {"type": "pip", "requirements_files": None},
-                r"requirements_files\n  none is not an allowed value",
+                r"none is not an allowed value",
             ),
             (
                 {"type": "pip", "requirements_build_files": None},
-                r"requirements_build_files\n  none is not an allowed value",
+                r"none is not an allowed value",
             ),
         ],
     )
     def test_invalid_packages(self, input_data: dict[str, Any], expect_error: str) -> None:
-        with pytest.raises(pydantic.ValidationError, match=expect_error):
-            # doesn't pass type check: https://github.com/pydantic/pydantic/issues/1847
-            pydantic.parse_obj_as(PackageInput, input_data)  # type: ignore
+        with pytest.raises((pydantic.ValidationError, TypeError), match=expect_error):
+            adapter = pydantic.TypeAdapter(PackageInput)
+            adapter.validate_python(input_data)
```

Comment:
When does it raise TypeError? It shouldn't raise TypeError :thinking:

Reply:
As far as I know Pydantic 2 catches some errors and not others by default because it wants to give users the flexibility to handle validation errors in their own way. It catches only ValueError, Assertion and custom ValidationError.

If I changed TypeError to for example KeyError or RuntimeError, still the same result.
Thats why I had change it [here](https://github.com/containerbuildsystem/cachi2/pull/328/files#diff-4b875352bcb272a1d35c2b56db986573fc40bd9c2a1d0d008c9d15c1270dd3a0) as well, because of:

```
@staticmethod
def _validate(value: Any) -> "RootedPath":
    if not isinstance(value, (str, PathLike)):
        raise TypeError(f"expected str or os.PathLike, got {type(value).__name__}")
    return RootedPath(path=value)
```

But maybe there is some workaround here :man_artist:

Reply:
Ah, I see. Let's raise ValueError instead then